spec/formatting.md

```diff
@@ -197,18 +197,21 @@ the following steps are taken:
 
 1. If the _expression_ includes an _operand_, resolve its value.
    If this fails, use a _fallback value_ for the _expression_.
-2. Based on the _function_ starting sigil and _name_,
+2. _Resolve the name_ of the _function_ and, based on the starting sigil,
    find the appropriate function implementation from the _function registry_.
    If the registry does not define an implementation for this _name_,
    emit an Unknown Function error
    and use a _fallback value_ for the _expression_.
-3. Resolve the _option_ values to a mapping of string identifiers to values.
+3. If the _expression_ includes _options_, resolve the _options_ to a mapping
+   of string identifiers to values.
    For each _option_:
-   - If its right-hand side successfully resolves to a value,
-     bind the _name_ of the _option_ to the resolved value in the mapping.
+   - _Resolve the name_ of the _option_.
+   - If the _option_'s _name-part_ already exists in the resolved mapping of _options_,
+     emit a Duplicate Option Name error.
+   - If the _option_'s right-hand side successfully resolves to a value,
+     bind the _name-part_ of the _option_'s to the resolved value in the mapping.
```

Comment:
Is this saying that formatting an expression `{:func one:foo=42 two:foo=13}` would generate an error? That would be very surprising, as we'd effectively be saying that the `ns:` part of an option name is completely meaningless.

Reply:
See: the description of this PR 😀. It does currently mean exactly this. 

I don't think it makes it meaningless, but much depends on how namespaced things (currently only functions and options) are implemented. If the `ns:` says, effectively, where to find the registry entry for validation but not, for example, the invocation, then it probably *is* an error to have `one:foo`/`two:foo` (you can only call one function and that one either expects the `foo` from `one:` or the `foo` from `two:`, but not both. This also alleviates the function from having to figure out that `one:foo` means "my version of `foo`". If the namespace is programmable, the called function has no way of knowing what it is.

Is there a function that would expect two separately-namespaced options of the otherwise-same name?

Reply:
> If the namespace is programmable, the called function has no way of knowing what it is.

I think that *function* namespaces and *option* namespaces should be controlled by different actors, because they serve different purposes.

* Function namespaces are defined **by i18n engineers** who integrate an MF2 implementation into a codebase. They help disambiguate between different packages providing formatters of the same name.
* Option namespaces are defined **by implementors of add-ons** to already existing functions. For example, `:datetime ns:skeleton=YYY` means that the implementation of `datetime` conforms to the default registry, and on top of that, it offers an extra option, `ns:skeleton`.

Note that functions themselves are a kind of a namespace for their options: In both `ns:func foo=bar`, `foo` isn't spelled as `ns:foo` because it's already effectively namespaced by the fact of being an option to `ns:func`. It's the add-on options that need namespacing, and that namespacing should not be programmable.

> Is there a function that would expect two separately-namespaced options of the otherwise-same name? (and, yes, I know this is different from the normal behavior of "namespaces")

As you note, two separately namespaced options are unlikely, because ultimately, there's only one implementation being invoked anyways.

However, we should consider how namespacing impacts other *unnamespaced* options. The namespacing of add-on options guarantees forward compatibility, in case the function being extended adds a new option in the future.

This is why functions should be able to see the namespaces of all their options.

Reply:
In other words: we probably don't need `one:foo=value` and `two:foo=value` to be different, but we do need `foo=value` and `one:foo=value` to be different because o forward compatibility of the add-on mechanism.

The easiest way to achieve it is to pass options together with their namespaces to the function during invocation. This in turn makes `one:foo=value` and `two:foo=value` different options.

Reply:
I agree with @stasm. I would expect that the `:func` implementation to get called with an options bag
```js
{ 'one:foo': '42', 'two:foo': '13' }
```

Reply:
Done, although I added text permitting an implementation to do processing of the options.